src/editor/dag.rs

```diff
@@ -702,6 +703,61 @@ impl<'arena, Node: Ast<'arena>> Dag<'arena, Node> {
         self.write_text(&mut s, format);
         s
     }
+
+    ///  Generate `.dot` code of the internal state of the `Dag`
+    pub fn to_dot_code(&self) -> String {
+        /// function that helps to connect nodes with edges
+        fn add_to_graph<'arena, Node: Ast<'arena>>(
+            node: &'arena Node,
+            digraph_edges: &mut String,
+            hmap_nodes: &mut HashMap<usize, String>,
+        ) {
+            let name = node as *const Node as usize;
+            hmap_nodes.entry(name).or_insert(format!(
+                "node{} [label=\"{}\"]\n",
+                name,
+                node.debug_name()
+            ));
+
+            if node.children().len() > 0 {
+                for &child in node.children() {
+                    add_to_graph(child, digraph_edges, hmap_nodes);
+                }
+
+                let edge = format!("node{} -> ", name);
+                digraph_edges.push_str(&edge);
+
+                for i in 0..node.children().len() {
+                    let child_name = node.children()[i] as *const Node as usize;
+                    if i != node.children().len() - 1 {
+                        digraph_edges.push_str(&format!("node{}, ", child_name));
+                    } else {
+                        digraph_edges.push_str(&format!("node{}", child_name));
+                    }
+                }
+                digraph_edges.push_str("\n");
+            }
+        }
+
+        let mut hmap_nodes = HashMap::<usize, String>::new();
+        let mut digraph_head = "digraph G {\n".to_owned();
+        let mut digraph_edges = "\n".to_owned();
+        let digraph_tail = "\n}";
+
+        for snapshot in &self.root_history {
+            add_to_graph(snapshot.root, &mut digraph_edges, &mut hmap_nodes);
+        }
+        for value in hmap_nodes.values() {
+            digraph_head.push_str(&value);
+        }
+        digraph_head.push_str("\n");
+
+        digraph_head.push_str(&digraph_edges);
+        digraph_head.push_str(&digraph_tail);
+
+        log::debug!("\n\n{} \n", digraph_head);
```

Comment:
I'm not sure we need to debug this in here; I think it's probably better to let the caller decide what to do with the output. 

I also suspect that the dot code will get very big very fast, so I can imagine situations where someone would turn on debug printing to narrow down some unrelated issue and then their log file gets filled with thousands of lines of dot code that isn't relevant to what they're doing.

Reply:
Ops, I forgot to delete this.